### ISSUE TITLE
Lab 6 level 4

### ISSUE THREAD

**Issue by rickyaandrew** [nus-cs2030/2021-s1#206]
Hiii guyss may i ask how the flatmap is being implemented? is it possible to do it without using @SuppressWarnings("unchecked")?

currently i am using that, but i just curious if that is the only way to implement it. any suggestions or tips?

**Comment by aaron-ljx** [comment]
Yes, I did it without using  SuppressWarnings("unchecked").  If you're using that, it suggests that you're not casting properly. Maybe you can look into generics again and see where you went wrong.

**Comment by marcuspeh** [comment]
I did it without supress warning. Looks the the warning and the generics part, probs somewhere went wrong.

**Comment by SkyBlaise99** [comment]
Would just like to ask if you guys used `get()` for `flatMap`. Cuz `function.apply(thing)` gives `Logger<? extends U>` and I would like to make it into `LoggerImpl<? extends U>` or even better, just `LoggerImpl<U>`

**Comment by aaron-ljx** [comment]
Does LoggerImpl<U<U>> logger = (LoggerImpl<U<U>>)func.apply(value);  work for you?

**Comment by SkyBlaise99** [comment]
There's the type cast safety thing, won't this fail codecrunch?
`[Java 16777761] [W] Type safety: Unchecked cast from capture#2-of ? extends Logger<? extends U> to LoggerImpl<U>` <- this is from my compiler

**Comment by aaron-ljx** [comment]
This is my flatmap method signature: 
public <U<U>> Logger <U<U>> flatMap(Function<? super T, Logger<U<U>> func)
Since LoggerImpl is an instance of Logger, there should be no problem returning LoggerImpl at all. 
By the way, I didn't get any type cast safety warning though when doing that. Perhaps it might have something to do with my flatmap method signature.

**Comment by SkyBlaise99** [comment]
i used `flatMap(Function<? super T, ? extends Logger<? extends U>> f)`. This is the method signiture given by codecrunch. Did you get an error for using `Logger<U>`? If not then i guess i change my code lol
![Screenshot (40)](https://user-images.githubusercontent.com/69448615/96334894-b0b1f400-10a6-11eb-9c5b-754403eac7ec.png)

**Comment by aaron-ljx** [comment]
Nope, I didn't get any error for that.

**Comment by SkyBlaise99** [comment]
So i guess it is allowed? But why codecrunch gives such a weird signiture lol
actually on a second thought, since i got A alr, might as well not change the thing haha

**Comment by rickyaandrew** [comment]
> This is my flatmap method signature:
> public <U> Logger <U> flatMap(Function<? super T, Logger<U> func)
> Since LoggerImpl is an instance of Logger, there should be no problem returning LoggerImpl at all.
> By the way, I didn't get any type cast safety warning though when doing that. Perhaps it might have something to do with my flatmap method signature.

ohh yeaahh i also used `<U> Logger <U> flatMap(Function<? super T, Logger<U> func)`, does it matter? i still passed all the test cases thooo, but not sure whats the difference right there between `<U> Logger <U> flatMap(Function<? super T, Logger<U> func)` and `<U> Logger <? extends U> flatMap(Function<? super T, Logger<? extends U> func)`

**Comment by aaron-ljx** [comment]
> 
> 
> > This is my flatmap method signature:
> > public  Logger  flatMap(Function<? super T, Logger func)
> > Since LoggerImpl is an instance of Logger, there should be no problem returning LoggerImpl at all.
> > By the way, I didn't get any type cast safety warning though when doing that. Perhaps it might have something to do with my flatmap method signature.
> 
> ohh yeaahh i also used `<U> Logger <U> flatMap(Function<? super T, Logger<U> func)`, does it matter? i still passed all the test cases thooo, but not sure whats the difference right there between `<U> Logger <U> flatMap(Function<? super T, Logger<U> func)` and `<U> Logger <? extends U> flatMap(Function<? super T, Logger<? extends U> func)`

Since ? extends U means all child classes of U and U, I think that it's there to make the code more flexible than just class U.

**Comment by charoi** [comment]
Just use **var** haha it works

**Comment by Maurice2n97** [comment]
Use Logger<? extends U> instead of Logger<U> as the variable type to store the variable you get from f.apply(value)
